include/snap/metric.h

```diff
@@ -12,35 +12,152 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-#ifndef METRIC_H
-#define METRIC_H
+#ifndef SNAP_METRIC_METRIC_H
+#define SNAP_METRIC_METRIC_H
 
-#include <ctime>
+#include <chrono>
 #include <map>
 #include <string>
 #include <vector>
 
-namespace Metric {
+#include "plugin.pb.h"
 
-class Namespace : private std::vector<std::string> {
-};
+#include "snap/config.h"
+
+namespace Plugin {
 
-// TODO (danielscottt): use a union for data here.
-class Metric {
+/**
+ * A PODS describing an element in a metric namespace.
+ */
+class Metric final {
 
   public:
-    void setData(int data);
+
+    struct NamespaceElement {
+      public:
+        /**
+         * value is the static value of this node in a namespace.
+         * When a namespace element is _not_ dynamic, value is used. During
+         * metric collection, value should contain the static name for this metric.
+         */
+        std::string value;
+
+        /**
+         * name is used to describe what this dynamic element is querying against.
+         * E.g. in the namespace `/intel/kvm/[vm_id]/cpu_wait` the element at index
+         * 2 has the name "vm_id".
+         * @see value
+         */
+        std::string name;
+
+        /**
+         * description is the description of this namespace element.
+         */
+        std::string description;
+    };
+
+    Metric();
+    /**
+     * The typical metric constructor.
+     * @param ns The metric's namespace.
+     * @param unit The metric's unit.
+     * @param description The metric's description.
+     */
+    Metric(std::vector<NamespaceElement> ns, std::string unit,
+           std::string description);
+
+    /**
+     * This constructor is used in the collector proxy during collect metrics.
+     * It's used to wrap the rpc::Metric type with the metric type from this
+     * library.
+     */
+   Metric(rpc::Metric* metric);
+
+    Metric(const Metric& from);
+
+    ~Metric();
+
+    /**
+     * dynamic_ns_elements returns the indices in the metric's namespace which
+     * are dynamic.
+     */
+    std::vector<int> dynamic_ns_elements();
+
+    /**
+     * ns returns the metric's namespace.
+     * If there is a memoized copy, that is returned. Else the namespace is
+     * copied into the cache then returned.
+     */
+    const std::vector<NamespaceElement>& ns();
+
+    /**
+     * set_ns sets the namespace of the metric in its `rpc::Metric` ptr.
+     * It also invalidates the memoization cache of the namespace if it is
+     * present.
+     * @see memo_ns
+     */
+    void set_ns(std::vector<NamespaceElement>);
```

Comment:
Word -- So this is an antipattern?  I assumed things like `enum`s, `class`es, `struct`s, & c. were supposed to be `PascalCase`, and everything else `snake_case`.